pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,438 @@
+	}
+
+	recoveryPodManifestPath := path.Join(s.PodManifestDir, RecoveryPodFileName)
+	err = ioutil.WriteFile(recoveryPodManifestPath, recoveryPodBytes, 644)
+	if err != nil {
+		return fmt.Errorf("failed to write recovery pod manifest %q: %v", recoveryPodManifestPath, err)
+	}
+
+	// Create client cert
+	ca := crypto.CA{
+		Config:          rootCaConfig,
+		SerialGenerator: &crypto.RandomSerialGenerator{},
+	}
+
+	// Create client certificates for system:admin
+	clientCert, err := ca.MakeClientCertificateForDuration(&user.DefaultInfo{Name: "system:admin"}, certValidity)
+	if err != nil {
+		return fmt.Errorf("failed to create client certificate: %v", err)
+	}
+
+	clientCertBytes, clientKeyBytes, err := clientCert.GetPEMBytes()
+
+	s.restConfig = &rest.Config{
+		Host: "https://localhost:7443",
+		TLSClientConfig: rest.TLSClientConfig{
+			CAFile:   servingCaCertPath,
+			CertData: clientCertBytes,
+			KeyData:  clientKeyBytes,
+		},
+	}
+
+	kubeconfig, err := s.KubeConfig()
+	if err != nil {
+		return fmt.Errorf("failed to create kubeconfig: %v", err)
+	}
+
+	kubeconfigBytes, err := yaml.Marshal(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to marshal kubeconfig: %v", err)
+	}
+
+	kubeconfigPath := path.Join(recoveryDirPath, AdminKubeconfigFileName)
+	err = ioutil.WriteFile(kubeconfigPath, kubeconfigBytes, 600)
+	if err != nil {
+		return fmt.Errorf("failed to write kubeconfig %q: %v", kubeconfigPath, err)
+	}
+
+	return nil
+}
+
+func (s *Apiserver) WaitForHealthz(ctx context.Context) error {
+	// We don't have to worry about connecting to an old, terminating apiserver
+	// as our client certs are unique to this instance
+
+	kubeClientset, err := s.GetKubeClientset()
+	if err != nil {
+		return err
+	}
+
+	return wait.PollUntil(500*time.Millisecond, func() (bool, error) {
+		req := kubeClientset.RESTClient().Get().AbsPath("/healthz")
+
+		klog.V(1).Infof("Waiting for recovery apiserver to come up at %q", req.URL())
+
+		_, err = req.DoRaw()
+		if err != nil {
+			klog.V(5).Infof("apiserver returned error: %v", err)
+			return false, nil
+		}
+
+		return true, nil
+	}, ctx.Done())
+}
+
+func (s *Apiserver) Destroy() error {
+	recoveryPodManifestPath := path.Join(s.PodManifestDir, RecoveryPodFileName)
+	err := os.Remove(recoveryPodManifestPath)
+	if err != nil {
+		return fmt.Errorf("failed to remove recovery pod manifest %q: %v", recoveryPodManifestPath, err)
+	}
+
+	// TODO: consider removing the recoveryDir
+	// (The dir is tmp by default anyways and putting a wrong path here could be fatal as it would have to recursively delete.)
+
+	return nil
+}
+
+func copyFile(src, dest string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("failed to open file %q: %v", src, err)
+	}
+	defer srcFile.Close()
+
+	srcFileInfo, err := srcFile.Stat()
+	if err != nil {
+		return fmt.Errorf("failed to stat file %q: %v", src, err)
+	}
+
+	if srcFileInfo.IsDir() {
```

Comment:
check for IsRegular instead. There are other non-files than dirs.

Reply:
```golang
// A FileInfo describes a file and is returned by Stat and Lstat.
type FileInfo interface {
	Name() string       // base name of the file
	Size() int64        // length in bytes for regular files; system-dependent for others
	Mode() FileMode     // file mode bits
	ModTime() time.Time // modification time
	IsDir() bool        // abbreviation for Mode().IsDir()
	Sys() interface{}   // underlying data source (can return nil)
}
```

also symlinks are fine I guess